### PR TITLE
feature/MAGICDRAW-489

### PR DESCRIPTION
**Pull request by w3e**

### CODE REVIEW

**Comment by w3e** [comment]
Fixed issue and refactored code to use single class.

File: src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/SRConfigurator.java [hunk 1]
```diff
@@ -20,8 +20,10 @@
 public class SRConfigurator implements BrowserContextAMConfigurator, DiagramContextAMConfigurator {
 
     public static final String NAME = "Systems Reasoner";
-
-    private SRAction validateAction = null, createSpecificAction = null, ontoBehaviorAction = null, instance2BSTAction = null, createInstanceMenuAction = null, aspectAction, copyAction = null;
+    public static final String ID_NONRECURSIVE = "Specialize Structure";
+    public static final String ID_RECURSIVE = "Specialize Structure Recursive";
```
Comment (ivan-gomes) [inline src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/SRConfigurator.java]: *Recursively

File: src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/SRConfigurator.java [hunk 2]
```diff
@@ -157,7 +163,10 @@ public ActionsCategory handleSingleNode(ActionsCategory category, ActionsManager
             final Classifier classifier = (Classifier) element;
             validateAction = new ValidateAction(classifier);
             ontoBehaviorAction = new CreateOntoBehaviorBlocks(classifier, false);
-            createSpecificAction = new CreateSpecificAction(classifier, false);
+
+            specializeStructureAction = new SpecializeStructuresAction(classifier, false, ID_NONRECURSIVE, false, false);
```
Comment (ivan-gomes) [inline src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/SRConfigurator.java]: *ID

File: src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/SRConfigurator.java [hunk 3]
```diff
@@ -157,7 +163,10 @@ public ActionsCategory handleSingleNode(ActionsCategory category, ActionsManager
...
+            specializeStructureRecursiveAction = new SpecializeStructuresAction(classifier, false, ID_RECURSIVE, true, false);
+            createBSTAction = new SpecializeStructuresAction(classifier, false, ID_RECURSIVE_INDIVIDUALLY, true, true);
```
Comment (ivan-gomes) [inline src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/SRConfigurator.java]: *ID_RECURSIVE_INDIVIDUAL

File: src/main/java/gov/nasa/jpl/mbee/mdk/validation/actions/CreateSpecializedTypeAction.java [hunk 4]
```diff
@@ -26,95 +24,141 @@
...
     private String name;
+    private boolean isIndividual;
+
 
-    public CreateSpecializedTypeAction(final Property property, final Classifier parent) {
-        this(property, parent, true, DEFAULT_NAME);
-    }
 
-    public CreateSpecializedTypeAction(final Property property, final Classifier parent, final boolean redefineAttributes, final String name) {
+    public CreateSpecializedTypeAction(final Property property, final Classifier parent, final boolean redefineAttributes, final String name, boolean isIndividualß) {
```
Comment (ivan-gomes) [inline src/main/java/gov/nasa/jpl/mbee/mdk/validation/actions/CreateSpecializedTypeAction.java]: *isIndividual

File: src/main/java/gov/nasa/jpl/mbee/mdk/validation/actions/CreateSpecializedTypeAction.java [hunk 5]
```diff
@@ -26,95 +24,141 @@
...
     }
 
-    public static final Classifier createSpecializedClassifier(final Classifier general, final Classifier parent, final Property property) {
+    private static Type getExistingSpecial(StructuralFeature structuralFeature) {
+        Set<Type> types = new HashSet<Type>();
+        Element owner = structuralFeature.getOwner();
+        for(RedefinableElement redef : structuralFeature.getRedefinedElement()){
+            if(redef instanceof TypedElement){
+                types.add(((TypedElement) redef).getType());
+                System.out.println("Found type: "+((TypedElement) redef).getType().getName() +" id  "+ ((TypedElement) redef).getType().toString() + " for SF " + structuralFeature.getName() + "  " + structuralFeature.toString());
```
Comment (ivan-gomes) [inline src/main/java/gov/nasa/jpl/mbee/mdk/validation/actions/CreateSpecializedTypeAction.java]: Remove debugging.

File: src/main/java/gov/nasa/jpl/mbee/mdk/validation/actions/CreateSpecializedTypeAction.java [hunk 6]
```diff
@@ -26,95 +24,141 @@
...
+            }
+        }
+        for(Element oe : owner.getOwnedElement()){
+            if(oe instanceof Property){
+                if(!oe.equals(structuralFeature)){
+                for(RedefinableElement redef : ((Property) oe).getRedefinedElement()){
+                    if(redef instanceof TypedElement) {
+                        if (types.contains(((TypedElement) redef).getType())) {
+                            System.out.println("Found type: "+((TypedElement) oe).getType().getName() +" id  "+ ((TypedElement) oe).getType().toString() + " for SF " + redef.getName() + "  " + redef.toString());
```
Comment (ivan-gomes) [inline src/main/java/gov/nasa/jpl/mbee/mdk/validation/actions/CreateSpecializedTypeAction.java]: Removing debugging.

File: src/main/java/gov/nasa/jpl/mbee/mdk/validation/actions/CreateSpecializedTypeAction.java [hunk 7]
```diff
@@ -26,95 +24,141 @@
...
+
+        return null;
+    }
+
+    public static final Classifier createSpecializedClassifier(final Classifier general, final Classifier parent, final StructuralFeature structuralFeature) {
         for (final Class<? extends Classifier> c : UNSPECIALIZABLE_CLASSIFIERS) {
             if (c.isAssignableFrom(general.getClass())) {
                 Application.getInstance().getGUILog()
-                        .log("Warning: " + (property != null ? property.getQualifiedName() : "< >") + " is a " + c.getSimpleName() + ", which is not specializable.");
+                        .log("Warning: " + (structuralFeature != null ? structuralFeature.getQualifiedName() : "< >") + " is a " + c.getSimpleName() + ", which is not specializable.");
```
Comment (ivan-gomes) [inline src/main/java/gov/nasa/jpl/mbee/mdk/validation/actions/CreateSpecializedTypeAction.java]: *[WARNING]

File: src/main/java/gov/nasa/jpl/mbee/mdk/validation/actions/CreateSpecializedTypeAction.java [hunk 8]
```diff
@@ -26,95 +24,141 @@
...
-        return special;
+        Classifier specific = (Classifier) CopyPasting.copyPasteElement(general, parent);
+
+        ArrayList<NamedElement> members = new ArrayList<>();
+        for(NamedElement ne : specific.getOwnedMember()){
+            members.add(ne);
+        }
+         for(NamedElement member : members){
+            if(member instanceof RedefinableElement) {
+                System.out.println(member.getClassType().getName() + " removing " + member.getName());
```
Comment (ivan-gomes) [inline src/main/java/gov/nasa/jpl/mbee/mdk/validation/actions/CreateSpecializedTypeAction.java]: Removing debugging.

File: src/main/java/gov/nasa/jpl/mbee/mdk/validation/actions/RedefineAttributeAction.java [hunk 9]
```diff
@@ -18,32 +18,35 @@
...
     private Classifier subClassifier;
     private RedefinableElement re;
-    private boolean createSpecializedType;
+    private boolean createSpecial;
     private String name;
+    private boolean isIndividual;
 
-    public RedefineAttributeAction(final Classifier clazz, final RedefinableElement re) {
-        this(clazz, re, false, DEFAULT_NAME);
+    public RedefineAttributeAction(final Classifier clazz, final RedefinableElement re, boolean isIndividually) {
```
Comment (ivan-gomes) [inline src/main/java/gov/nasa/jpl/mbee/mdk/validation/actions/RedefineAttributeAction.java]: *isIndividual

File: src/main/java/gov/nasa/jpl/mbee/mdk/validation/actions/RedefineAttributeAction.java [hunk 10]
```diff
@@ -18,32 +18,35 @@
...
      * @param name
+     * @param isIndividually
      */
-    public RedefineAttributeAction(final Classifier targetForRedefEl, final RedefinableElement elementToBeRedefined, final boolean createSpecializedType, final String name) {
+    public RedefineAttributeAction(final Classifier targetForRedefEl, final RedefinableElement elementToBeRedefined, final boolean createSpecializedType, final String name, boolean isIndividually) {
         super(name);
         this.subClassifier = targetForRedefEl;
         this.re = elementToBeRedefined;
-        this.createSpecializedType = createSpecializedType;
+        this.createSpecial = createSpecializedType;
```
Comment (ivan-gomes) [inline src/main/java/gov/nasa/jpl/mbee/mdk/validation/actions/RedefineAttributeAction.java]: Pick either createSpecial or createSpecializedType.

File: src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/actions/SpecializeStructuresAction.java [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+import gov.nasa.jpl.mbee.mdk.lib.Utils;
+import gov.nasa.jpl.mbee.mdk.validation.actions.AddInheritanceToAssociationAction;
+import gov.nasa.jpl.mbee.mdk.validation.actions.RedefineAttributeAction;
+
+import java.awt.*;
+import java.awt.event.ActionEvent;
+import java.util.ArrayList;
+import java.util.List;
+
+public class SpecializeStructuresAction extends SRAction {
```
Comment (ivan-gomes) [inline src/main/java/gov/nasa/jpl/mbee/mdk/systems_reasoner/actions/SpecializeStructuresAction.java]: *SpecializeStructureAction

**Comment by w3e** [comment]
Added requested changes.

File: src/main/java/gov/nasa/jpl/mbee/mdk/validation/actions/CreateSpecializedTypeAction.java [hunk 12]
```diff
@@ -126,7 +125,7 @@ public static final Classifier createSpecializedClassifier(final Classifier gene
         for (final Class<? extends Classifier> c : UNSPECIALIZABLE_CLASSIFIERS) {
             if (c.isAssignableFrom(general.getClass())) {
                 Application.getInstance().getGUILog()
-                        .log("Warning: " + (structuralFeature != null ? structuralFeature.getQualifiedName() : "< >") + " is a " + c.getSimpleName() + ", which is not specializable.");
+                        .log("[WARNING]: " + (structuralFeature != null ? structuralFeature.getQualifiedName() : "< >") + " is a " + c.getSimpleName() + ", which is not specializable.");
```
Comment (ivan-gomes) [inline src/main/java/gov/nasa/jpl/mbee/mdk/validation/actions/CreateSpecializedTypeAction.java]: No colon. I'll change it for you.

**Pull request merged**